src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
 
     Node const *stack[64];
     Node const **stack_ptr = stack;
-    *stack_ptr++ = nullptr;
-    Node const *node = _bvh.getRoot();
+    *stack_ptr++ = nullptr;            // push sentinel value
+    Node const *node = _bvh.getRoot(); // start with root
     do
     {
-      Node const *child_left = _bvh.getNodePtr(node->left_child);
-      Node const *child_right = _bvh.getNodePtr(node->right_child);
+      int left_child_index = node->left_child;
+      int right_child_index = node->right_child;
+
+      Node const *left_child_node = _bvh.getNodePtr(left_child_index);
+      Node const *right_child_node = _bvh.getNodePtr(right_child_index);
+
+      bool left_child_is_leaf = left_child_node->isLeaf();
+      bool right_child_is_leaf = right_child_node->isLeaf();
+
+      bool prune_left =
+          (left_child_is_leaf && pruneLeaf(queryIndex, left_child_index)) ||
+          (!left_child_is_leaf &&
+           pruneLeftSubtree(queryIndex, left_child_index));
```

Comment:
Yeah, I don't like it, but this is currently the only way. Ideally, the hierarchy would know whether something is a leaf or not based on its index, and you would simply need a single `prune` thing.

Reply:
Another comment I have is that there's no particular reason to call the function `pruneLeftSubtree`, just call it `pruneSubtree`. The reason is that the function itself is only dependent on the indices of query and node, not on whether that node is a left or right subtree.

Reply:
I disagree. The meaning of what we prune is derived from the context. The name of the function should be based on what it does, which is actually just a check that has nothing to do with whether it is a left or a right subtree. But I am not going to require a change.